### PR TITLE
security: harden API endpoints and client session management

### PR DESCRIPTION
**Pull request by Joker5514**
- Disable X-Powered-By header in server.js and api/handler.js to avoid
  leaking the Express version to potential attackers.
- Add rate limiters (20 req/min) to /api/license/validate and
  /api/license/status, which were previously unrestricted.
- Validate usageEvent (alphanumeric string, max 128 chars) and units
  (positive integer, max 100 000) in /api/usage/record to prevent log
  injection and oversized payloads from authenticated clients.
- Add source-type check in validateLicenseToken so only Stripe-issued
  tokens (source === 'stripe') are accepted; auth tokens can no longer
  be cross-used on license endpoints.
- Add 30-day max-age expiry to client-side loadSession() so stale
  sessions stored in localStorage are automatically invalidated.

https://claude.ai/code/session_01Rw7pnL5CcfDtHHBvFBqidN

## Release Notes

* **New Features**
  * Added license validation and status management endpoints with rate limiting.
  * Implemented 30-day session expiration for enhanced account security.
  * Added BSRNN vocal separation model support for advanced audio processing.
  * Enhanced validation for usage event recording.

* **Security**
  * Disabled X-Powered-By header to prevent version exposure in responses.

### CODE REVIEW

**Comment by Joker5514** [comment]
## 🤖 Randy Code Agent — Automated Review

**Rating:** REQUEST_CHANGES

## Summary

This PR adds several hardening changes: disables Express `X-Powered-By`, rate-limits license validation/status endpoints, tightens license token validation to only accept Stripe-issued tokens, bounds usage event/unit inputs, and expires client-side sessions after 30 days. It also includes unrelated ML pipeline changes that add BSRNN to default models, move worker patch/cache setup after the message handler, and change the worker init payload shape.

## Potential Bugs or Logic Errors

### Blocking: existing license tokens may be invalidated

`api/monetization.js` now rejects any token whose payload does not include `source: 'stripe'`. New tokens include that field, but previously issued valid license tokens appear likely to lack it, so this can immediately invalidate existing customers’ licenses across `/license/validate`, `/license/status`, `/license/activate`, `/usage/record`, and authenticated checkout flows.

Recommendation: add a migration/compatibility path, e.g. accept legacy tokens until their natural expiry, version tokens with a `typ`/`aud` claim going forward, or provide an explicit reissue flow before enforcing `source === 'stripe'`.

### Possible entitlement bypass / behavior change in worker init

`public/app/pipeline-orchestrator.js` now sends `allowedStages: 32` and `allowedModels: ['vad', 'rnnoise', 'demucs', 'bsrnn']` unconditionally. If stage/model access is meant to be tier-gated, this appears to bypass the client-side capability limits defined in `auth.js` (`FREE` max 16, `PRO` max 24, etc.).

Recommendation: derive `allowedStages` and `allowedModels` from the authenticated user/session caps rather than hard-coding the maximum.

### Rate limits may behave incorrectly behind proxies

The in-memory limiter keys on `req.ip`. If the app runs behind a proxy/CDN/load balancer and Express `trust proxy` is not configured, many users may share the same internal IP and be rate-limited together. Conversely, if proxy trust is configured incorrectly, spoofed IP headers can bypass limits.

Recommendation: explicitly configure proxy trust for the deployment environment and/or use a battle-tested limiter with a shared store.

## Code Quality and Best Practices Issues

- The PR mixes security/API hardening with unrelated ML pipeline/model changes. Splitting these into separate PRs would make review and rollback much safer.
- The custom in-memory rate limiter is simple and readable, but it will reset on process restart and will not coordinate across multiple instances/serverless workers.
- The session expiration constant is a good improvement, but `loadSession()` should also reject invalid/non-numeric/future `loginTime` values rather than only checking missing/too-old timestamps.
- The BSRNN integration assumes a `[1, numBins]` tensor and output names like `vocal_mask` or `output`; without tests or model contract checks this is likely brittle and may silently fall back via the `catch`.

## Security Concerns

- Requiring a Stripe/license-flow marker on license JWTs is directionally good, but using only `source` is weaker than standard JWT claims. Prefer validating `aud`, `iss`, `typ`, and possibly token version to prevent token confusion.
- Client-side session tier validation remains advisory only; any privileged/paid capabilities must be enforced server-side as well.
- Rate-limiting unauthenticated validation/status endpoints is positive, but per-IP unauthenticated limits can become a denial-of-service vector for users sharing an IP.

## Suggestions for Improvement

1. Add legacy-token compatibility or a planned reissue/migration flow before enforcing `payload.source === 'stripe'`.
2. Split ML worker/orchestrator changes into a separate PR from API/session security changes.
3. Add tests for:
   - legacy license token validation behavior,
   - new `source: 'stripe'` token validation,
   - usage event/unit validation edge cases,
   - session expiry including missing, stale, invalid, and future `loginTime`,
   - tier-derived worker `allowedStages`/`allowedModels`.
4. Replace or wrap the custom limiter with a standard library and shared backing store if this app can run on more than one instance.
5. Use stronger JWT semantics (`iss`, `aud`, `typ`, `ver`) instead of relying only on a custom `source` field.

## Overall Rating

**REQUEST_CHANGES** — the hardening direction is good, but the likely invalidation of existing license tokens and the hard-coded maximum worker entitlements should be addressed before merging.

File: api/monetization.js [hunk 1]
```diff
@@ -180,6 +180,8 @@ function validateLicenseToken(token) {
     if (!crypto.timingSafeEqual(expected, got)) return null;
     const payload = JSON.parse(Buffer.from(parts[1], 'base64url').toString());
     if (Date.now() / 1000 > payload.exp) return null;
+    // Only accept tokens issued by the Stripe/license flow, not auth tokens
+    if (payload.source !== 'stripe') return null;
     return payload;
```
[the remark this answers is not in the document]
Comment (Joker5514) [reply api/monetization.js:185]: @copilot apply changes based on [this feedback](https://github.com/Joker5514/VoiceIsolate-Pro/pull/464#discussion_r3195026283)

File: public/app/pipeline-orchestrator.js [hunk 2]
```diff
@@ -414,20 +350,87 @@ class PipelineOrchestrator {
...
+      this.mlWorker.postMessage({
+        type:    'init',
+        payload: {
+          modelBasePath:      '/app/models/',
+          preferredProviders: ['webgpu', 'wasm'],
+          // Canonical four — matches model-loader.js + models-manifest.json.
+          // ml-worker.js MODEL_FILES maps these aliases to .onnx filenames.
+          allowedModels:      ['vad', 'rnnoise', 'demucs', 'bsrnn'],
+          allowedStages:      32,
+        },
```
[the remark this answers is not in the document]
Comment (Joker5514) [reply public/app/pipeline-orchestrator.js:455]: @copilot apply changes based on [this feedback](https://github.com/Joker5514/VoiceIsolate-Pro/pull/464#discussion_r3195026310)

File: api/monetization.js [hunk 3]
```diff
@@ -213,6 +215,8 @@ function makeRateLimiter(maxReqs, windowMs) {
 
 const activateLimiter    = makeRateLimiter(5,  60_000); // 5 activations per minute per IP
 const usageRecordLimiter = makeRateLimiter(30, 60_000); // 30 usage records per minute per IP
+const validateLimiter    = makeRateLimiter(20, 60_000); // 20 validations per minute per IP
+const statusLimiter      = makeRateLimiter(20, 60_000); // 20 status checks per minute per IP
```
[the remark this answers is not in the document]
Comment (Joker5514) [reply api/monetization.js:219]: @copilot apply changes based on [this feedback](https://github.com/Joker5514/VoiceIsolate-Pro/pull/464#discussion_r3195026322)

[the remark this answers is not in the document]
Comment (Joker5514) [reply api/monetization.js:219]: @copilot apply changes based on [this feedback](https://github.com/Joker5514/VoiceIsolate-Pro/pull/464#discussion_r3195026322)

File: api/monetization.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Joker5514) [reply api/monetization.js:185]: @copilot apply changes based on [this feedback](https://github.com/Joker5514/VoiceIsolate-Pro/pull/464#discussion_r3195026344)
